File: frontend/lib/auth/session-cookie.ts
import { createHmac, timingSafeEqual } from 'crypto';

type Role = 'user' | 'partner' | 'admin' | 'super_admin';
const VALID_ROLES = new Set<Role>(['user', 'partner', 'admin', 'super_admin']);

function cookieSecret(): string {
  return (
    process.env.SESSION_COOKIE_SECRET
    || process.env.OTP_PEPPER
    || (process.env.NODE_ENV === 'production' ? '' : 'dev-session-cookie-secret')
  );
}

function decodePayload(token: string): string | null {
  try {
    const pad = '='.repeat((4 - (token.length % 4)) % 4);
    const b64 = token.replace(/-/g, '+').replace(/_/g, '/') + pad;
    return Buffer.from(b64, 'base64').toString('utf8');
  } catch {
    return null;
  }
}

export function verifySessionCookie(token: string | undefined | null): { uid: string; role: Role } | null {
  if (!token) return null;
  const secret = cookieSecret();
  if (!secret) return null;

  const raw = decodePayload(token);
  if (!raw) return null;

  const parts = raw.split('|');
  if (parts.length !== 4) return null;
  const [uid, role, expStr, sig] = parts;
  if (!VALID_ROLES.has(role as Role)) return null;

  const payload = `${uid}|${role}|${expStr}`;
  const expected = createHmac('sha256', secret).update(payload).digest('hex');
  const sigBuf = Buffer.from(sig, 'utf8');
  const expBuf = Buffer.from(expected, 'utf8');
  if (sigBuf.length !== expBuf.length || !timingSafeEqual(sigBuf, expBuf)) return null;

  const exp = Number(expStr);
  if (!Number.isFinite(exp) || exp < Math.floor(Date.now() / 1000)) return null;

  return { uid, role: role as Role };
}

export class LoginOtpRequiredError extends Error {
  constructor() {
    super('login_otp_required');
    this.name = 'LoginOtpRequiredError';
  }
}

export async function syncSessionCookie(idToken: string): Promise<Role | null> {
  // Go through the same-origin /api rewrite rather than hitting the backend
  // directly. A direct call to NEXT_PUBLIC_API_URL is cross-origin from the
  // page, so the CSP connect-src ('self' + supabase) blocks it before the
  // request is even sent — which surfaces as a bare "Failed to fetch".
  // The rewrite in next.config.js forwards /api/* to the backend, and
  // app/api/auth/session/route.ts still wins for its own path because
  // filesystem routes take precedence over afterFiles rewrites.
  const res = await fetch('/api/auth/session-token', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ id_token: idToken }),
  });
  if (!res.ok) {
    let detail = '';
    try {
      const body = (await res.json()) as { detail?: string };
      detail = body.detail ?? '';
    } catch {
      /* ignore */
    }
    if (res.status === 403 && detail === 'login_otp_required') {
      throw new LoginOtpRequiredError();
    }
    return null;
  }
  const data = (await res.json()) as { token?: string; role?: Role };
  if (!data.token || !data.role) return null;

  await fetch('/api/auth/session', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ token: data.token }),
  });

  return data.role;
}

export async function clearSessionCookie(): Promise<void> {
  await fetch('/api/auth/session', { method: 'DELETE' });
}

File: backend/core/session_cookie.py
"""Signed session cookies for Next.js middleware (HttpOnly, not forgeable)."""
from __future__ import annotations

import base64
import hashlib
import hmac
import time

from core.config import settings

VALID_ROLES = frozenset({"user", "partner", "admin", "executive", "super_admin"})
DEFAULT_TTL_SECONDS = 86_400


def _secret() -> bytes:
    raw = settings.SESSION_COOKIE_SECRET or settings.OTP_PEPPER or settings.RESEND_API_KEY
    if not raw:
        if settings.is_production:
            raise RuntimeError("SESSION_COOKIE_SECRET (or OTP_PEPPER) is required in production")
        raw = "dev-session-cookie-secret"
    return raw.encode()


def sign_session(uid: str, role: str, *, ttl_seconds: int = DEFAULT_TTL_SECONDS) -> str:
    if role not in VALID_ROLES:
        raise ValueError(f"Invalid role: {role}")
    exp = int(time.time()) + ttl_seconds
    payload = f"{uid}|{role}|{exp}"
    sig = hmac.new(_secret(), payload.encode(), hashlib.sha256).hexdigest()
    raw = f"{payload}|{sig}"
    return base64.urlsafe_b64encode(raw.encode()).decode().rstrip("=")


def verify_session(token: str) -> dict[str, str] | None:
    try:
        pad = "=" * (-len(token) % 4)
        raw = base64.urlsafe_b64decode(token + pad).decode()
        uid, role, exp_str, sig = raw.rsplit("|", 3)
        payload = f"{uid}|{role}|{exp_str}"
        expected = hmac.new(_secret(), payload.encode(), hashlib.sha256).hexdigest()
        if not hmac.compare_digest(sig, expected):
            return None
        if int(exp_str) < time.time():
            return None
        if role not in VALID_ROLES:
            return None
        return {"uid": uid, "role": role}
    except Exception:
        return None
